### DISCUSSION TITLE
Pattern Matching should respect [DoesNotReturn] and [DoesNotReturnIf]

### DISCUSSION THREAD

**Discussion by dazerdude** [dotnet/csharplang#7795] [Language Ideas] [answered]
This (somewhat unexpectedly for me) produces a compiler error.

```
public void Read(ref Utf8JsonReader reader) {
    var node = JsonNode.Parse(ref reader);
    Assert(node is JsonObject obj);
    Console.WriteLine(obj.ToString());  // CS0165: Use of unassigned local variable `obj`
}

public void Read(ref Utf8JsonReader reader) {
    var node = JsonNode.Parse(ref reader);
    if (node is not JsonObject obj) { Fail(); }
    Console.WriteLine(obj.ToString());  // CS0165: Use of unassigned local variable `obj`
}

[DoesNotReturn]
private static void Fail() => throw new Exception();

private static void Assert([DoesNotReturnIf(false)] bool cond) {
    if (cond) { throw new Exception(); }
}
```

whereas using nearly-equivalent null checks works as expected

```
public void Read(ref Utf8JsonReader reader) {
    var obj = JsonNode.Parse(ref reader) as JsonObject;
    Assert(obj != null);
    Console.WriteLine(obj.ToString());
}

public void Read(ref Utf8JsonReader reader) {
    var obj = JsonNode.Parse(ref reader) as JsonObject;
    if (obj == null) { Fail(); }
    Console.WriteLine(obj.ToString());
}
```

I assume this is because `[DoesNotReturn]` and `[DoesNotReturnIf]` is only taken to account when analyzing nullability, but it seems like it should apply to certain other scenarios as well since it declares pretty general code reachability information. Shouldn't the compiler be able to use the attribute information to determine that the final line is only reachable if `obj` is populated, or am I missing some bit of subtlety here?

**Comment by KennethHoff** [comment]
Obj is only assigned in a debug context, so this wouldn't be possible regardless.

**Reply by dazerdude** [reply]
That's particular to `Debug.Assert` and isn't really relevant to my point here. I was just using it as convenient example, but the issue described is more general than that. I've updated to the sample to hopefully provide some clarity.

**Reply by dazerdude** [reply]
Weirdly this compiles for me in Release w/o nullable static analysis complaining:

```
public void Read(ref Utf8JsonReader reader)
{
    object? obj = null;
    Debug.Assert(obj != null);
    Console.WriteLine(obj.ToString()); // throws NullPointerException at runtime
}
```

whereas this produces the expected error in Release:

```
public void Read(ref Utf8JsonReader reader)
{
    object? obj = null;
#if DEBUG
    Debug.Assert(obj != null);
#endif
    Console.WriteLine(obj.ToString()); // CS8602: Dereference of a possibly null reference.
}
```

But I'm probably misunderstanding some detail of `Debug.Assert`. I don't use it much in my own code.

**Comment by TahirAhmadov** [comment]
The problem here is that "variable assignment" is a core language compilation error, and they don't want to tie it to certain attributes. There are proposals to make "does not return" a core language feature, which would allow your scenarios.

**Reply by dazerdude** [reply]
Huh, that makes sense I guess. I suppose I was thinking that nullable reference types and the related static analysis were also core language features of modern c# (especially given the amount of attention it gets).

Just seems odd to me for it to have different behavior/requirements from other language features, even if it is optional.

**Reply by TahirAhmadov** [reply]
NRT static analysis is a little strange in that it is not a core language feature technically - it is run by built-in analyzers - but those analyzers are built and supported by Microsoft.

**Reply by dazerdude** [reply]
Right. I can see how that might be a distinction that matters from Microsoft's point of view, but from the outside the implementation details of the various static analysis features are both opaque and awkward justification for a behavioral difference like this. It shouldn't _really_ matter to the language that the NRT feature is statically analyzed by a built-in Roslyn analyzer and uninitialized variables are statically analyzed by the compiler. From my point of view, I just run the compiler and it generates errors.

I feel a bit like Mr. Incredible
![tenor](https://github.com/dotnet/csharplang/assets/463403/869fb8d3-69ca-4617-958e-51f09c2584ad)

**Reply by TahirAhmadov** [reply]
I think variable assignment analysis is different from the NRT analysis in that 1) NRT analysis was only recently added as a huge breaking change using conditions like preprocessor directives and project settings to introduce it gradually, which means it's difficult to make it a "core language feature", 2) NRT analysis is by definition imperfect - there are pending requests to improve it but even then there are situations when `!` is necessary - whereas the variable assignment can and has to be determined perfectly - otherwise situations can occur which make C# "unsafe" when developer expects "safety", which absolutely cannot happen (as opposed to potential NREs).

**Reply by CyrusNajmabadi** [reply]
> are both opaque and awkward justification 

I don't think it's awkward. Indeed, I think asking for the language to be impacted by attributes is more awkward. 

If flow analysis is to change, it should be within the language. 

Note: nrt really is the special "odd man out" case. It is awkward and strange precisely because it's trying to graft a change in semantics over billions of lives of code that are 20+ years old. 

Nrt is the last resort.  And we should not do more in its style unless absolutely necessary. 

Flow analysis isn't suffering with the same issues that nrt did, so the solution doesn't need the same compromises

**Reply by dazerdude** [reply]
> I think variable assignment analysis is different from the NRT analysis in that 1) NRT analysis was only recently added as a huge breaking change using conditions like preprocessor directives and project settings to introduce it gradually, which means it's difficult to make it a "core language feature", 2) NRT analysis is by definition imperfect - there are pending requests to improve it but even then there are situations when `!` is necessary - whereas the variable assignment can and has to be determined perfectly - otherwise situations can occur which make C# "unsafe" when developer expects "safety", which absolutely cannot happen (as opposed to potential NREs).

These are different features, yes. But under the hood I would expect them to be performing the same static analysis w.r.t branched reachability of statements. Those reachability implementations currently differ from each other, and while your points are valid for why NRT is (and should remain) optional, I don't see how they impact reachability in specific. Why should it matter which feature is leveraging the reachability analyzer? The compiler should be able to consistently determine which statements are reachable and which are not.

NRT as a feature is incomplete and should remain optional, but the reachability extensions it uses don't appear to be incomplete in the same way, and more generally useful? This is why I scoped this topic to these to attributes in specific and not remainder of NRT attributes. `[DoesNotReturn]` and `[DoesNotReturnIf]` are enforced in the implementation and should be safe to use for reachability analysis generally unless I'm missing something?

Now that being said, I can understand why MSFT wouldn't want to use the existing Attributes to support doing this, and I do appreciate your comments to this point. They have been enlightening. Is there some issue I'm not seeing with promoting them to a keyword (or tbh automatically annotating methods with this information if they do not return (`[DoesNotReturnIf]` would be harder to do automatically, but I think possibly still doable)).

**Reply by CyrusNajmabadi** [reply]
> These are different features, yes. But under the hood I would expect them to be performing the same static analysis

They are definitely not. Nrt is weak heuristics, with all sorts of special cases to avoid a deluge of false positives. 

Assignment analysis is precise, and will not allow a single false negative. 

They are about as far away as one can get and they serve entirely different purposes based on entirely different histories.  

Importantly, assignment analysis isn't trying to make up for mistakes of the past. It's been a general success, and we don't have any desire to change the core principles that it was based on.

**Reply by TahirAhmadov** [reply]
> Is there some issue I'm not seeing with promoting them to a keyword

Yes but I don't remember which one - likely a discussion, but I'm not sure. Look for "return type never" or something to that effect.

**Reply by TahirAhmadov** [reply]
> Importantly, assignment analysis isn't trying to make up for mistakes of the past. It's been a general success, and we dummy have any desire to change the core principles that it was based on.

Wouldn't both assignment analysis and NRT analysis be affected if/when "return type never" is implemented?

**Reply by CyrusNajmabadi** [reply]
> Why should it matter which feature is leveraging the reachability analyzer? The compiler should be able to consistently determine which statements are reachable and which are not.

This is def not hte case.  If we did the same analysis for both then we'd either end up with:

1. NRT analysis being an awful experience, with 100-1000x more false positives that would make the feature generally useless for users.
2. reachability analysis having the same problems of NRT where the system thinks the code is fine, but you can still blow up at runtime.

Both of these are unpalatable to us.

**Reply by CyrusNajmabadi** [reply]
>  Is there some issue I'm not seeing with promoting them to a keyword

I see no issue with that.  Fundamentally though, it's not about it being a keyword or not.  It's about it being *enforced* or not.  If something had such a keyword/attribute, we'd have to require that it be properly implemented in order to have the lang take a dependency on it.  Having a method state it does not return, but still then return would not be acceptable as it would form a safety hole.

**Reply by CyrusNajmabadi** [reply]
> Wouldn't both assignment analysis and NRT analysis be affected if/when "return type never" is implemented?

Likely yes?

**Reply by TahirAhmadov** [reply]
>  It's about it being enforced or not.

Correct me if I'm wrong, the enforcement would be done by a "core rule", not an analyzer, and stylistically it's much more likely that a "core rule" is driven by a keyword, and not an attribute, right?
Meaning, a "core rule" can be driven off of an attribute, but at that point it would be "weird", right?

**Reply by dazerdude** [reply]
(Oh my god, so many comments, I'm so slow. I wrote this immediately after my previous comment. I should've refreshed I guess.)

 >  I don't think it's awkward. Indeed, I think asking for the language to be impacted by attributes is more awkward.

I merely meant that I find implementation details of the compiler to be an awkward justification for contract inconsistencies between the two implementations of flow analysis. (As an aside, I find this generally true of most software, implementation details influencing contracts is a reality of the world we live in, but a disappointing one). After @TahirAhmadov's comments, I don't disagree that using attributes would probably be the wrong approach for a more general extension to reachability, and totally understand why it is currently this way.

 > Flow analysis isn't suffering with the same issues that NRT did, so the solution doesn't need the same compromises

Well, yes and no. I agree that this wouldn't be something I'd complain about if NRT had never been implemented. But I'd argue that all other features that rely on flow analysis suffer by proxy. Because NRT's implementation of flow analysis has additional features in it's flow analysis, I don't think it's unreasonable for a programmer to expect those flow analysis extensions to be supported by other features that are built on top of flow analysis. Those extensions are more generally applicable than merely to NRT (if only for programmer comfort).

(I'm using flow analysis and reachability analysis interchangeably here. Please correct me if that's wrong.)

**Reply by CyrusNajmabadi** [reply, marked as answer]
@dazerdude The way to think about this is that these are different systems in the language design, with different goals.  I understand your desire for them to be unified an to have uniform impact in the language feature-set. 

All i can say is: that was considered, and explicitly rejected because the resultant language would actually have more negatives than positives over what we have today.  

This is often a fundamental tension we have in lang design.  Often times there is a preference for purity and uniformity.  However, this butts heads with the need for pragmatic outcomes.  NRT is fundamentally solving a different set of issues in a different fashion, due to many complex factors.  Trying to do the same with definite assignment is a non-goal, and we feel would actually result in a worse situation.

**Reply by dazerdude** [reply]
> Reachability analysis having the same problems of NRT where the system thinks the code is fine, but you can still blow up at runtime.

> Having a method state it does not return, but still then return would not be acceptable as it would form a safety hole.

Would it blow up at runtime? I totally get the need for adequate enforcement of reachability is absolutely necessary and that NRT in general is full of holes. But I'm under the impression that `[DoesNotReturn]` or `[DoesNotReturnIf]` in specific were adequately enforced. They appear to have produced all the necessary errors in annotated methods in my experience at least. Is there a known issue w/ NRT's flow analysis w.r.t. these attributes?

Unless I'm mistaken the other NRT attributes don't really effect statement reachability the same way. I completely understand that those shouldn't be applied to other features.

**Reply by dazerdude** [reply]
> @dazerdude The way to think about this is that these are different systems in the language design, with different goals. I understand your desire for them to be unified an to have uniform impact in the language feature-set.
> 
> All i can say is: that was considered, and explicitly rejected because the resultant language would actually have more negatives than positives over what we have today.
> 
> This is often a fundamental tension we have in lang design. Often times there is a preference for purity and uniformity. However, this butts heads with the need for pragmatic outcomes. NRT is fundamentally solving a different set of issues in a different fashion, due to many complex factors. Trying to do the same with definite assignment is a non-goal, and we feel would actually result in a worse situation.

The tension between purity and pragmatism is fundamental to _software_ and I completely understand. Thanks for the discussion. I do really appreciate the insights you and @TahirAhmadov have provided me here.

**Reply by CyrusNajmabadi** [reply]
> Would it blow up at runtime? 

Yes.  Consider something as simple as:

```c#
int i;
if (whatever)
{
    DoesNotReturn();
}
else
{
     i = ComputeActualValue();
}

Use(i);
```

Now imagine that the code that provides `DoesNotReturn` actually has the IL equivalent to `return 0;`.  Now what?  If we just blindly trusted that method, then we can literally get to the point that we're `Use`ing `i` despite it not being initialized.  This could lead to all sorts of blow-ups or security issues.

**Reply by CyrusNajmabadi** [reply]
> But I'm under the impression that [DoesNotReturn] or [DoesNotReturnIf] in specific were adequately enforced.
> They appear to have produced all the necessary errors in annotated methods in my experience at least.

THere is no actual enforcement.  The following compiles successfully:

```c#
    [DoesNotReturn]
    public T DoesNotReturn<T>()
    {
        return default(T);
    }
```

**Reply by CyrusNajmabadi** [reply]
>  do really appreciate the insights you and @TahirAhmadov have provided me here.

YW :)

**Reply by dazerdude** [reply]
> There is no actual enforcement. The following compiles successfully:
> ```cs
>     [DoesNotReturn]
>     public T DoesNotReturn<T>()
>     {
>         return default(T);
>     }
> ```

NRT enforces it and produces CS8763. This error isn't a nullability error, just a general flow analysis error. The only reason NRT is enforcing it appears to me to be an implementation detail. It _could_ be applied to other flow analysis features (like uninitialized variables) if it was enforced by a non-optional C# feature. Or is there something else going on w/ the IR that would prevent this?

>  If we just blindly trusted that method

We don't need to blindly trust the method call. We're trusting that the compiler adequately enforced the relevant errors (like CS8763) on the implementation side of the method call. Or is this not true for some reason? Is there some gap in the implementation side enforcement of `[DoesNotReturn]`? From my naive position it appears to not have very many edge cases.

**Reply by dazerdude** [reply]
Is the trick that we can't necessarily trust that the method we're calling because it could be in another dll and that dll could be malicious and designed to lie to our code (returning when it told us it wouldn't)? I could see that being an unacceptable security risk, but then you're calling into an untrusted binary already anyway?

**Reply by CyrusNajmabadi** [reply]
> NRT enforces it and produces CS8763. 

There is no enforcement.

```c#
1>ConsoleApp1 -> C:\Temp\collections\bin\Debug\net8.0\ConsoleApp1.dll
1>Done building project "ConsoleApp1.csproj".
```

This is a legal program that produces a legal assembly which has that attribute, but doesn't abide by it.

**Reply by CyrusNajmabadi** [reply]
> We're trusting that the compiler adequately enforced the relevant errors

But it literally does not.  I showed that above.  The above is a legal program, which states that the method does not return, when it actually does.  As such, "trusting" can and will lead to invalid memory accesses and security issues.  That's not acceptable.

**Reply by dazerdude** [reply]
I think you're misunderstanding my current point a little. I understand that the current compiler doesn't do this, and CS8763 is only produced with NRT enabled. But it doesn't have to be that way? If the compiler _did_ enforce CS8763 (or some equivalent error) always, then it _could_ also trust that methods marked non-returning weren't lying?

**Reply by CyrusNajmabadi** [reply]
> and that dll could be malicious

It's not even about it being malicious. I wrote a trivial example that compiles successfully and completely violates the attribute. 

It is simply not enforced and has no impact.

**Reply by CyrusNajmabadi** [reply]
> I think you're misunderstanding my current point a little. I understand that the current compiler doesn't do this, and CS8763 is only produced with NRT enabled. But it doesn't have to be that way? If the compiler _did_ enforce CS8763 always, then it _could_ also trust that methods marked non-returning weren't lying?

As mentioned:

I see no issue with that. Fundamentally though, it's not about it being a keyword or not. It's about it being enforced or not. If something had such a keyword/attribute, we'd have to require that it be properly implemented in order to have the lang take a dependency on it.

:-)

**Reply by dazerdude** [reply]
> I see no issue with that. Fundamentally though, it's not about it being a keyword or not. It's about it being enforced or not. If something had such a keyword/attribute, we'd have to require that it be properly implemented in order to have the lang take a dependency on it.
> 
> :-)

I must've missed that earlier in the flurry of replies. Yeah, I get it's not about the specific language construct, which is why I was trying to examine whether there was some issue in my logic, or if there was some known gap in the current errors NRT produces in relation to `[DoesNotReturn]`.

Anyway, thanks a ton. This has been enlightening. =]

**Reply by GabeSchaffer** [reply]
Given this program:
> ```cs
> int i;
> if (whatever)
> {
>     DoesNotReturn();
> }
> else
> {
>      i = ComputeActualValue();
> }
> 
> Use(i);
> ```

What if it were translated into this?
```cs
int i;
if (whatever)
{
    DoesNotReturn();
    ThrowHelper.ThrowInvalidProgramException();
}
else
{
     i = ComputeActualValue();
}
Use(i);
```
Of course this would break any program that relied on a `[DoesNotReturn]` method actually returning, but that seems like a fairly low risk.

Then conceivably the runtime could recognize `[DoesNotReturn]` attributes and either replace every `return` in such annotated methods with `throw` (or refuse to verify assemblies containing such methods that return), and elide the `ThrowInvalidProgramException()` calls when it inlines methods like the one above.

**Comment by TahirAhmadov** [comment]
https://github.com/dotnet/csharplang/issues/538
